From bbd6579e8644fbb0e4d61a712c9bb0533f6f077b Mon Sep 17 00:00:00 2001
From: llin4 <ling.lin@intel.com>
Date: Thu, 26 Jul 2012 12:59:30 +0800
Subject: system: Enable houdini feature

BZ: 23454

Houdini enable hook in system core to enable ARM application
read cpuinfo using cat

Change-Id: I734668b3dcfa34e8741e5214d172df68dd73c620
Signed-off-by: llin4 <ling.lin@intel.com>
Reviewed-on: http://android.intel.com:8080/58759
Reviewed-by: Wang, Zhenghua <zhenghua.wang@intel.com>
Reviewed-by: Beare, Bruce J <bruce.j.beare@intel.com>
Reviewed-by: Li, XiaojingX <xiaojingx.li@intel.com>
Tested-by: Li, XiaojingX <xiaojingx.li@intel.com>
Reviewed-by: buildbot <buildbot@intel.com>
Tested-by: buildbot <buildbot@intel.com>
---
 toolbox/Android.mk |    4 ++++
 toolbox/cat.c      |   25 +++++++++++++++++++++++++
 2 files changed, 29 insertions(+), 0 deletions(-)

diff --git a/toolbox/Android.mk b/toolbox/Android.mk
index 8aac1bc..2e8e017 100644
--- a/toolbox/Android.mk
+++ b/toolbox/Android.mk
@@ -1,6 +1,10 @@
 LOCAL_PATH:= $(call my-dir)
 include $(CLEAR_VARS)
 
+ifeq ($(INTEL_HOUDINI), true)
+    LOCAL_CFLAGS += -DWITH_HOUDINI
+endif
+
 TOOLS := \
 	ls \
 	mount \
diff --git a/toolbox/cat.c b/toolbox/cat.c
index 6ac31f8..ae0cb4f 100644
--- a/toolbox/cat.c
+++ b/toolbox/cat.c
@@ -183,6 +183,11 @@ raw_cat(int rfd)
 	}
 }
 
+#ifdef WITH_HOUDINI
+#define APP_WITH_ABI2 "/data/data/.appwithABI2"
+#define ARM_CPUINFO "/system/lib/arm/cpuinfo"
+#endif
+
 static void
 raw_args(char **argv)
 {
@@ -219,6 +224,26 @@ skipnomsg:
 				continue;
 			}
 			filename = *argv++;
+
+#ifdef WITH_HOUDINI
+			if (!strcmp(filename, "/proc/cpuinfo")) {
+				int uid_fd;
+				uid_t uid, my_uid;
+				uid_fd = open(APP_WITH_ABI2, O_RDONLY);
+				if (uid_fd != -1) {
+					my_uid = getuid();
+					while (read(uid_fd, &uid, 4) != 0) {
+						if (uid == my_uid) {
+							if (fd >= 0)
+								close(fd);
+							fd = open(ARM_CPUINFO, O_RDONLY);
+							break;
+						}
+					}
+					close(uid_fd);
+				}
+			}
+#endif
 		}
 		raw_cat(fd);
 		if (fd != fileno(stdin))
-- 
1.7.4.1

